### ISSUE TITLE
WCAG 2.4.3 A: Visually hidden navigation elements in keyboard focus order (Issue 24)

### ISSUE THREAD

**Issue by CJE001** [civictheme/uikit#368]
### Summary

Via Vision Australia assessment: August 2024

Impact: medium

Note: DTA have a 90 day remediation period to address the identified issues within the audit, all issues must be resolved to obtain WCAG 2.2 certification for digital.gov.au.

### Steps to reproduce

Global issue – Left side navigation
[Digital Experience | digital.gov.au](https://www.digital.gov.au/policy/digital-experience) (https://www.digital.gov.au/policy/digital-experience)

### Observed outcome

When the "Digital Experience Policy" navigation is expanded, keyboard focus moves through the visible second-level navigation link, and then through all of its visually hidden third-level navigation links. 

**Digital Experience Policy navigation - expanded state**
![Picture22](https://github.com/user-attachments/assets/c3918a6e-3b63-4f46-8821-54f42658e5e6)
 
**Focus moves through visually hidden third-level navigation links**
![Picture23](https://github.com/user-attachments/assets/618ee79a-96f6-4301-abed-ad5b2c83420c)

### Code used
```
<ul class="ct-menu ct-menu--level-0 ct-theme-light ct-side-navigation__menu " data-component-name="ct-menu">
<li class="ct-menu__item ct-menu__item--level-0 ct-menu__item--has-children ct-menu__item--active-trail " data-collapsible="true" data-collapsible-duration="0" aria-expanded="true">
<a class="ct-link ct-theme-light ct-menu__item__link" href="/policy/digital-experience" title="Digital Experience Policy" aria-expanded="true">
Digital Experience Policy
</a>
<a href="#" class="ct-menu__item__link-trigger" data-collapsible-trigger="" aria-expanded="true" title="Expand Digital Experience Policy menu">
<svg class="ct-icon ct-collapsible__icon" width="24" height="24" aria-hidden="true" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
...
</svg>
</a>
<div class="ct-menu__sub-menu__wrapper ct-menu__sub-menu__wrapper--level-1" data-collapsible-panel="" aria-hidden="false" style="">
<ul class="ct-menu ct-menu__sub-menu ct-menu--level-1">
<li class="ct-menu__item ct-menu__item--level-1 ct-menu__item--has-children " data-collapsible="true" data-collapsible-duration="0" data-collapsible-collapsed="" aria-expanded="false">
<a class="ct-link ct-theme-light ct-menu__item__link" href="/policy/digital-experience/digital-service-standard" title="Digital Service Standard" aria-expanded="false">
Digital Service Standard
</a>
<a href="#" class="ct-menu__item__link-trigger" data-collapsible-trigger="" aria-expanded="false" title="Expand Digital Service Standard menu">
<svg class="ct-icon ct-collapsible__icon" width="24" height="24" aria-hidden="true" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"></svg>
</a>
<div class="ct-menu__sub-menu__wrapper ct-menu__sub-menu__wrapper--level-2" data-collapsible-panel="" aria-hidden="true">
<ul class="ct-menu ct-menu__sub-menu ct-menu--level-2">
<li class="ct-menu__item ct-menu__item--level-2 ">
<a class="ct-link ct-theme-light ct-menu__item__link" href="/policy/digital-experience/digital-service-standard/services" title="Services covered by the Digital Service Standard">
Services covered by the Digital Service Standard
</a>
</li>
<li class="ct-menu__item ct-menu__item--level-2 ">
<a class="ct-link ct-theme-light ct-menu__item__link" href="/policy/digital-experience/digital-service-standard/dss-transition" title="Transition approach">
Transition approach
</a>
</li>
... 
</ul>
</div>
</li>
</ul>
</div>
</li>
</ul>
```

### Why this matters

Users who navigate via keyboard are forced to navigate through all focusable elements within the sub-navigation lists. This situation leads to unnecessary tab stops, which can be particularly challenging for users with limited dexterity, mobility disabilities, or those who may experience physical discomfort while navigating. 

As the elements are not visible, users may lose track of their location or experience difficulty determining where they are. 
Additionally, sighted screen reader users may become confused when their screen reader announces content that is not visible.

### Expected outcome

Ensure that when the sub-navigation lists are not visible (collapsed state), that they are accessibly hidden from all users. This can be achieved with the CSS `display:none`; property on the element that contains the navigation content. When the sub-navigation lists are displayed (expanded state) update the CSS to` display:block`;.

Additionally, the aria-hidden attribute must be removed from the parent `<div>` element for the sub-navigation, as outlined in Issue 33. 

The aria-expanded attribute must also be removed from the link that precedes the trigger for each sub-navigation list, as outlined in Issue 36.

Preferably, remove the word 'Expand' from the title attribute on the link element that triggers the sub-navigation. As this link has the aria-expanded attribute, its state will automatically be communicated to assistive technologies. 

The word 'Expand' may confuse some users when the element is in a collapsed state, and causes a redundant announcement when in an expanded state.

For example, the NDVA screen reader currently announces 'Expand Digital Experience Policy menu  collapsed link' and 'Expand Digital Experience Policy menu expanded link'. It is also recommended to not rely on the title attribute for the accessible name for these triggers, and instead use visually hidden text or an aria-label. 

Remove the title attribute from all page links, as the content of the link acts as the accessible name. When a link already has an accessible name. the title attribute will then act as the element description in the accessibility tree, which is resulting screen readers announcing duplicate information. 

Ideally, reimplement the sub-navigation trigger as a semantic `<button>` element. When screen reader users encounter a button, they expect it to trigger functionality (rather than navigate somewhere, as is the expectation of a link), making its purpose clearer. If this is not possible, the `role="button"` can be added to the link. 

### Code example

**HTML**

```
<nav aria-label="main">
<ul>
<li>
<a class="ct-link ct-theme-light ct-menu__item__link" href="/policy/digital-experience">
Digital Experience Policy
</a>
<button class="ct-menu__item__link-trigger" aria-expanded="true">
<span class="visually-hidden">
Digital experience policy menu
</span>
<svg class="ct-icon ct-collapsible__icon" width="24" height="24" focusable="false" aria-hidden="true" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
...
</svg>
</button>
<ul id="sub-menu1" style="display: block;">
<li class="ct-menu__item ct-menu__item--level-1 ct-menu__item--has-children " data-collapsible="true" data-collapsible-duration="0" data-collapsible-collapsed="">
<a class="ct-link ct-theme-light ct-menu__item__link" href="/policy/digital-experience/digital-service-standard">
Digital Service Standard
</a>
<button class="ct-menu__item__link-trigger" aria-expanded="false">
<span class="visually-hidden">
Digital Service Standard menu
</span>
<svg class="ct-icon ct-collapsible__icon" width="24" height="24" focusable="false" aria-hidden="true" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
...
</svg>
</button>
<div class="ct-menu__sub-menu__wrapper ct-menu__sub-menu__wrapper--level-2" data-collapsible-panel="" style="display:none;">
<ul class="ct-menu ct-menu__sub-menu ct-menu--level-2">
<li class="ct-menu__item ct-menu__item--level-2 ">
<a class="ct-link ct-theme-light ct-menu__item__link" href="/policy/digital-experience/digital-service-standard/services" title="Services covered by the Digital Service Standard">
Services covered by the Digital Service Standard
</a>
</li>
… 
</ul>
</div>
… 
</li>
</ul>
… 
</li>
</ul>
</nav>
```

**CSS**

```
.visually-hidden:not(:focus):not(:active) {
clip: rect(0 0 0 0);
clip-path: inset(50%);
height: 1px;
overflow: hidden;
position: absolute;
white-space: nowrap;
width: 1px;
}
```
### Related issues

For other issues related to this component, please refer to:

•	Issue 10 – Current page shown with presentation alone
•	Issue 22 – Sub-navigation links not keyboard accessible
•	Issue 23 – Illogical focus order when skip link activated
•	Issue 28 – Focused element not visible and has no visible focus indicator
•	Issue 33 – Element with aria-hidden attribute contains focusable descendants
•	Issue 36 – Link has invalid state

**Comment by alan-cole** [comment]
This ticket should be resolved with the fix in https://github.com/civictheme/uikit/pull/462

**Comment by daniel-grynenko** [comment]
**Test Env:** DEV
**Test Status:** PASS
**Test Result:** Verified that collapsed sub navigation elements are hidden from keyboard focus via the display: none; attribute. Confirmed that expanding the menu removes display: none; and correctly displays the sub navigation. Confirmed that aria-hidden attributes update correctly and do not conflict with display: none;.

'display: none' is present for collapsed sub navigation items
<img width="600" alt="Image" src="https://github.com/user-attachments/assets/f3ff9761-e94d-4fda-8570-9dc147b00c48" />

Using keyboard navigation to tab through side bar. Keyboard expectedly skips collapsed menu items

https://github.com/user-attachments/assets/5cface44-e28c-406c-a030-4f0bbb6ec6ff
